Readme.md
# Docker build


[![pipeline status](https://git.windmaker.net/a-castellano/DockerBuild/badges/master/pipeline.svg)](https://git.windmaker.net/a-castellano/DockerBuild/commits/master)[![coverage report](https://git.windmaker.net/a-castellano/DockerBuild/badges/master/coverage.svg)](https://git.windmaker.net/a-castellano/DockerBuild/commits/master)

An originary [Daedalus Project](https://github.com/daedalusproject) utility which allows users to create Docker images during CI/CD jobs.

This tool is curently used for creating all Docker images from [Limani](https://git.windmaker.net/a-castellano/limani) project. Images were only allowed to be uploaded to [Docker Hub](https://hub.docker.com/) but other registry URL's can be used.

## Installation

### From source

Just follow these instructions:
```bash
git clone https://git.windmaker.net/a-castellano/DockerBuild.git && cd DockerBuild
make build
sudo make install
```

### Using Repos

Add packages.windmaker.net repo:
```bash
apt-get install -y gnupg ca-certificates wget
wget -O - http://packages.windmaker.net/WINDMAKER-GPG-KEY.pub | sudo apt-key add - 
echo "deb [arch=amd64] http://packages.windmaker.net/ focal main" >> /etc/apt/sources.list 
apt-get update
apt-get install docker-build
```

## Usage

This tool uses environment variables in order to obtain registry name, credential, image names, etc.

The following environment variables must be set in order to make this tool work:

* **DOCKER_REGISTRY_USER** - specifies registry user.
* **DOCKER_REGISTRY_PASSWORD** - specifies registry user password.
* **DOCKER_IMAGES_MAINTAINER** - specifies Image maintainer name. 
* **IMAGE_NAME** - specifies the name of the image that this toll is creating. 
* **IS_BASE_IMAGE** - specifies is the image we want to create is a base image (see below)

Default registry is docker.io, when de default registry is being used the variable **DOCKER_ORGANIZATION_NAME** must be set, it can contain a user name instead of organizacion.

In order to use differente registry it mus be defined using variable **DOCKER_REGISTRY_URL**.

**docker-build** will look for a folder named as **IMAGE_NAME** value containing a valid Dockerfile.
```bash
export DOCKER_ORGANIZATION_NAME="acastellano"
export DOCKER_IMAGES_MAINTAINER="Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>"
export DOCKER_REGISTRY_USER="acastellano"
export DOCKER_REGISTRY_PASSWORD="my_dockerhub_password"
export IMAGE_NAME="test"
export IS_BASE_IMAGE=1

```
[![asciicast](https://asciinema.org/a/381553.svg)](https://asciinema.org/a/381553?speed=2)

## Base images

When [Limani](https://git.wind  ddmaker.net/a-castellano/limani) CI/CD creates docker images during master deploy, all docker images must use the same tag, base images generate a timestamp, non base images read that timetamps and will use it in order to generate the same tag as base image.


debian/changelog
docker-build (0.4-1) RELEASED; urgency=medium

  * Change Pakcage name
  * Allow the use of registry servers different form Docker Hub

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Mon, 28 Dec 2020 12:28:00 +0100

docker-build (0.2-4) RELEASED; urgency=medium

  * Minor fixes in CI

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Sun, 21 Feb 2020 11:13:00 +0100

docker-build (0.2-3) RELEASED; urgency=medium

  * Ready for production

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Wed, 01 May 2019 15:40:00 +0200

docker-build (0.2-2) UNRELEASED; urgency=medium

  * Remove some doblequotes

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Sun, 28 Apr 2019 00:40:36 +0200

docker-build (0.2-1) UNRELEASED; urgency=medium

  * Env variable usage fixed

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Sun, 27 Apr 2019 23:44:36 +0200

docker-build (0.1-6) UNRELEASED; urgency=medium

  * Docker build options have been fixed

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Sun, 27 Apr 2019 23:14:36 +0200

docker-build (0.1-5) UNRELEASED; urgency=medium

  * New test version

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Sun, 27 Apr 2019 22:50:36 +0200

docker-build (0.1-4) UNRELEASED; urgency=medium

  * Add more debug messages.

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Sun, 27 Apr 2019 22:30:36 +0200

docker-build (0.1-2) UNRELEASED; urgency=medium

  * Initial release.

 -- Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>  Sun, 27 Apr 2019 15:09:36 +0200


tests/01-debug.sh
#!/bin/bash -
#===============================================================================
#
#          FILE: 01-debug.sh
#
#   DESCRIPTION: Test debug

#        AUTHOR: Álvaro Castellano Vela (alvaro.castellano.vela@gmail.com)
#  ORGANIZATION: Daedalus Project
#       CREATED: 04/25/2019 07:26
#===============================================================================

source lib/build_docker_image_functions.sh

TearDown() {
    unset DEBUG
}

testDebugNoMessage() {
    no_menssage=$(debug "This is a test")
    assertEquals "" "$no_menssage" #If debug options is not enabled no message is printed
}

testDebugWithMessage() {
    DEBUG=1
    expectedMessage="This is a test"
    menssage=$(debug "$expectedMessage")
    assertSame "$expectedMessage" "$menssage" #If debug options is  enabled message is printed
}

# Load shUnit2.
. /usr/bin/shunit2


tests/02-variables.sh
#!/bin/bash -
#===============================================================================
#
#          FILE: 01-variables.sh
#
#   DESCRIPTION: Test varaible parsing

#        AUTHOR: Álvaro Castellano Vela (alvaro.castellano.vela@gmail.com)
#  ORGANIZATION: Daedalus Project
#       CREATED: 04/26/2019 05:58
#===============================================================================

source lib/build_docker_image_functions.sh

oneTimeSetUp() {
    export TMP_FOLDER="/var/tmp/dockerbuild/tests"
    mkdir -p $TMP_FOLDER
}

tearDown() {
    unset DOCKER_ORGANIZATION_NAME
    unset DOCKER_IMAGES_MAINTAINER
    unset DOCKERHUBUSER
    unset DOCKERHUBPASSWORD
    unset IMAGENAME
}

testNoVariables() {
    check_variables 2> $TMP_FOLDER/testNoVariables
    check_variables_status=$?
    check_variables_message=$(cat $TMP_FOLDER/testNoVariables)
    assertEquals "1" "$check_variables_status"
    assertEquals "Docker hub organization name not provided." "$check_variables_message"
}

testOnlyImagename() {
    IMAGENAME="SomeImage"
    check_variables 2> $TMP_FOLDER/testNoVariables
    check_variables_status=$?
    check_variables_message=$(cat $TMP_FOLDER/testNoVariables)
    assertEquals "1" "$check_variables_status"
    assertEquals "Docker hub organization name not provided." "$check_variables_message"
}

testWithDOCKER_ORGANIZATION_NAME() {
    DOCKER_ORGANIZATION_NAME="daedalusproject"
    check_variables 2> $TMP_FOLDER/testNoVariables
    check_variables_status=$?
    check_variables_message=$(cat $TMP_FOLDER/testNoVariables)
    assertEquals "1" "$check_variables_status"
    assertEquals "Image maintainer not provided." "$check_variables_message"
}

testWithDOCKER_IMAGES_MAINTAINER() {
    DOCKER_IMAGES_MAINTAINER="Álvaro Castellano Vela <alvaro.castellano.vela@gmail.com>"
    check_variables 2> $TMP_FOLDER/testNoVariables
    check_variables_status=$?
    check_variables_message=$(cat $TMP_FOLDER/testNoVariables)
    assertEquals "1" "$check_variables_status"
    assertEquals "Docker hub organization name not provided." "$check_variables_message"
}

# Load shUnit2.
. /usr/bin/shunit2


tests/01-test_debug.sh
#!/bin/bash -
#===============================================================================
#
#          FILE: 01-test_debug.sh
#
#   DESCRIPTION: Test debug

#        AUTHOR: Álvaro Castellano Vela (alvaro.castellano.vela@gmail.com)
#  ORGANIZATION: Daedalus Project
#       CREATED: 04/25/2019 07:26
#===============================================================================

source lib/build_docker_image_functions.sh

TearDown() {
	unset DEBUG
}

testDebugNoMessage() {
	no_menssage=$(debug "This is a test")
	assertEquals "" "$no_menssage" #If debug options is not enabled no message is printed
}

testDebugWithMessage() {
	DEBUG=1
	expectedMessage="This is a test"
	menssage=$(debug "$expectedMessage")
	assertSame "$expectedMessage" "$menssage" #If debug options is  enabled message is printed
}

# Load shUnit2.
. /usr/bin/shunit2


lib/build_docker_image_functions.sh
#!/bin/bash -
#===============================================================================
#
#          FILE: build_docker_image_functions.sh
#
#         USAGE: source lib/build_docker_image
#
#   DESCRIPTION: Build docker images.
#
#       OPTIONS: ---
#  REQUIREMENTS: docker
#          BUGS: ---
#         NOTES: ---
#        AUTHOR: Álvaro Castellano Vela (alvaro.castellano.vela@gmail.com)
#  ORGANIZATION: Daedalus Project
#       CREATED: 24/04/2019 06:59
#      REVISION: 0.2
#===============================================================================

# Functions

function debug {

	if [ -n "${DEBUG}" ]; then
		echo "$@"
	fi
}

function show_error {

	echo "$@" 1>&2
}

function check_variables {

	debug "Initiating check_variables."

	debug "Cheking first if registry URL has been set."

	if [[ -z ${DOCKER_REGISTRY_URL} ]]; then
		debug "DOCKER_REGISTRY_URL variable has not been set, asuming Dokcer Hub as registry."
		debug "Checking Docker Organization variables."
		if [[ -z ${DOCKER_ORGANIZATION_NAME} ]]; then
			show_error "Docker hub organization name (DOCKER_ORGANIZATION_NAME) not provided."
			return 1
		fi
	fi

	debug "Checking maintainer variable."
	if [[ -z ${DOCKER_IMAGES_MAINTAINER} ]]; then
		show_error "Image maintainer (DOCKER_IMAGES_MAINTAINER) not provided."
		return 1
	fi

	debug "Checking Docker login variables."

	if [[ -z ${DOCKER_REGISTRY_USER} ]]; then
		show_error "Registry user (DOCKER_REGISTRY_USER) not provided."
		return 1
	fi

	if [[ -z ${DOCKER_REGISTRY_PASSWORD} ]]; then
		show_error "Registry user password (DOCKER_REGISTRY_PASSWORD) not provided."
		return 1
	fi

	debug "Checking IMAGE_NAME variable."
	if [[ -z ${IMAGE_NAME} ]]; then

		show_error "IMAGE_NAME env variable is not set."
		return 1
	fi

	debug "Checking BUILD_PATH variable."
	if [[ -z ${BUILD_PATH} ]]; then

		BUILD_PATH="."
	fi
	return 0

}

function setup {

	debug "Initiating setup."

	if [[ ${IS_BASE_IMAGE} == 1 ]]; then

		debug "Current image is a base image type."
		mkdir -p ${BUILD_PATH}/timestamp
		DATESTRING=$(date +%Y%m%d%H%M)
		echo "${DATESTRING}" >${BUILD_PATH}/timestamp/timestampfile
	else
		debug "Current image is not a base image type."
		DATESTRING=$(cat ${BUILD_PATH}/timestamp/timestampfile)
		if [[ -z ${DATESTRING} ]]; then

			show_error "DATESTRING env variable is not set."
			return 1
		fi

		BASEIMAGE=$(grep "ARG BASE_IMAGE=" "${BUILD_PATH}/${IMAGE_NAME}"/Dockerfile | sed 's/:.*//' | sed 's/ARG BASE_IMAGE=//')
		debug "Checking if base image value is rightfully set."
		if [[ "${BASEIMAGE}" == "" ]]; then
			show_error "Cannot parse BASEIMAGE value."
			return 1
		fi
	fi

	debug "Setting build options."

	if [[ -z ${BUILD_OPTIONS} ]]; then
		BUILD_OPTIONS="--network=host --no-cache"
	fi

	debug "Setting tag options."

	if [[ -z ${LATEST_TAG_NAME} ]]; then
		LATEST_TAG_NAME="latest"
	fi

	debug "Determining Repository current branch name."

	if [[ -z ${CI_COMMIT_REF_NAME} ]]; then
		branch_name="$(git symbolic-ref HEAD 2>/dev/null)" || branch_name="(unnamed branch)"
		BRANCH=${branch_name##refs/heads/}
	else
		BRANCH=${CI_COMMIT_REF_NAME}
	fi

}

function rename_organization {

	if [[ ! -z ${DOCKER_REGISTRY_URL} ]]; then
		DOCKER_ORGANIZATION_NAME="${DOCKER_REGISTRY_URL}"
	fi
}

function build_image {

	debug "docker login"
	docker login --username "${DOCKER_REGISTRY_USER}" --password "${DOCKER_REGISTRY_PASSWORD}" ${DOCKER_REGISTRY_URL}

	docker build ${BUILD_OPTIONS} -t "${IMAGE_NAME}" -f "${IMAGE_NAME}/Dockerfile" .
	debug docker create --name="${IMAGE_NAME}" -i "${IMAGE_NAME}"
	docker create --name="${IMAGE_NAME}" -i "${IMAGE_NAME}"
}

function tag_image {

	debug docker tag "${IMAGE_NAME}" "${DOCKER_ORGANIZATION_NAME}"/"${IMAGE_NAME}":"${LATEST_TAG_NAME}"
	docker tag "${IMAGE_NAME}" "${DOCKER_ORGANIZATION_NAME}"/"${IMAGE_NAME}":"${LATEST_TAG_NAME}"

	if [ "${BRANCH}" == "master" ]; then
		debug docker tag "${IMAGE_NAME}" "${DOCKER_ORGANIZATION_NAME}"/"${IMAGE_NAME}":"${DATESTRING}"
		docker tag "${IMAGE_NAME}" "${DOCKER_ORGANIZATION_NAME}"/"${IMAGE_NAME}":"${DATESTRING}"
	fi
}

function commit_image {

	RELEASE_DATE=$(date)
	debug docker commit -m "${RELEASE_DATE}" -a "${DOCKER_IMAGES_MAINTAINER}" "${IMAGE_NAME}" "${DOCKER_ORGANIZATION_NAME}"/"${IMAGE_NAME}"
	docker commit -m "${RELEASE_DATE}" -a "${DOCKER_IMAGES_MAINTAINER}" "${IMAGE_NAME}" "${DOCKER_ORGANIZATION_NAME}"/"${IMAGE_NAME}"
	if [[ "${BRANCH}" == "master" ]]; then
		debug docker push "${DOCKER_ORGANIZATION_NAME}"/"${IMAGE_NAME}":"${DATESTRING}"
		docker push "${DOCKER_ORGANIZATION_NAME}"/"${IMAGE_NAME}":"${DATESTRING}"
	fi
	docker push "${DOCKER_ORGANIZATION_NAME}"/"${IMAGE_NAME}":latest
}

function clean_images {
	debug docker stop "${IMAGE_NAME}"
	docker stop "${IMAGE_NAME}"
	debug docker rm "${IMAGE_NAME}"
	docker rm "${IMAGE_NAME}"
	debug docker rmi "${IMAGE_NAME}"
	docker rmi "${IMAGE_NAME}"
}
